### PR TITLE
ocamlPackages.dose3: init 6.1

### PR DESCRIPTION
**Pull request by superherointj**
ocamlPackages.dose3: init 6.1

* Tests are failing. It needs further investigation, ideally from someone that better understands this library.

### CODE REVIEW

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}:
+
+buildDunePackage rec {
+  pname = "dose3";
+  version = "6.1";
+
+  src = fetchFromGitLab {
+    owner = "irill";
+    repo = "dose3";
+    rev = "8b3cc4285c96b97e379545a2b0718582d1d477cf";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dose3/default.nix:14]:
```suggestion
    rev = version;
```

Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Done.

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    repo = "dose3";
+    rev = version;
+    sha256 = "sha256-RFChY7VH2pVD4u5U1qng46h9aAv8I/3yXVaNmFDmKFI=";
+  };
+
+  minimalOCamlVersion = "4.03";
+  useDune2 = true;
+
+  buildInputs = [
+    ounit
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dose3/default.nix:22]: This is not needed.

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  ];
+  propagatedBuildInputs = [
+    bz2
+    camlzip
+    ocaml_extlib
+    ocamlgraph
+    stdlib-shims
+  ];
+
+  nativeCheckInputs = [
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dose3/default.nix:36]: What is this `nativeCheckInputs` attribute? Looks like dead code.

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    python39                  # Replaces: conf-python-3
+    python39Packages.pyyaml   # Replaces: conf-python3-yaml
+    dpkg                      # Replaces: conf-dpkg
+  ];
+  checkInputs = [
+    odoc
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dose3/default.nix:42]: This is not needed.

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    ounit
+  ];
+  doChecks = true;
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dose3/default.nix:45]:
```suggestion
  doCheck = true;
```

Also you may want to enable tests only when `ounit` is available (i.e., `lib.versionAtLeast ocaml.version "4.04"`).

**Review by vbgl — COMMENTED** [review]
There is some pending TODO in the `cudf` derivation.

File: pkgs/development/ocaml-modules/dose3/default.nix (hunk 5 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Thanks. Now it breaks on [tests](https://gitlab.com/irill/dose3/-/blob/master/src/applications/dune):
`dose-tests alias src/applications/runtest (exit 127)`, will investigate it.

File: pkgs/development/ocaml-modules/dose3/default.nix (hunk 4 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Removed

File: pkgs/development/ocaml-modules/dose3/default.nix (hunk 3 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Removed

File: pkgs/development/ocaml-modules/dose3/default.nix (hunk 2 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Removed

File: pkgs/development/ocaml-modules/dose3/default.nix (hunk 5 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: > There is some pending TODO in the `cudf` derivation.

Just enabled test for `cudf`. Thanks.

Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: I have extended the passthru.tests to other dependencies. I believe these tests are not run when building, only on CI. So it is less of problem to do this. As always let me know if I did something stupid. :-)

And yes, tests are still failing. =/

Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Current errors:
```
runTest (__main__.DoseTests)
Test : exitcode_distcheck_64_notexistent
Group : distcheck
Cmd : ./src/applications/distcheck.exe --checkonly a -f -e cudf://notexistent
Expected file : /dev/null
Expected exitcode : 64
 ... Fatal error in module dose_common.input: 
 Input file notexistent does not exist.
```
---
```
runTest (__main__.DoseTests)
Test : exitcode_distcheck_64_parse_error
Group : distcheck
Cmd : ./src/applications/distcheck.exe --checkonly a -f -e cudf://tests/DebianPackages/Packages_20201226T144309Z_amd64.xz
Expected file : /dev/null
Expected exitcode : 64
 ... Fatal error in module doseparse.stdLoaders: 
 Error while parsing CUDF from tests/DebianPackages/Packages_20201226T144309Z_amd64.xz (line: 1): unexpected RFC 822 token
```

Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: Is it possible to skip these?

File: pkgs/development/ocaml-modules/stdlib-shims/default.nix [hunk 6]
```diff
@@ -10,6 +10,9 @@ buildDunePackage rec {
   useDune2 = lib.versionAtLeast ocaml.version "4.08";
   minimumOCamlVersion = "4.02";
   doCheck = true;
+
+  passthru.tests = { inherit dose3; };
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/stdlib-shims/default.nix:14]: Can you elaborate on the meaning/purpose/motivation of this line?

Comment (superherointj) [reply pkgs/development/ocaml-modules/stdlib-shims/default.nix:14]: When this package is upgraded, I expect it to test dose3 package as well.

Comment (vbgl) [reply pkgs/development/ocaml-modules/stdlib-shims/default.nix:14]: I think that `nixpkgs-review` already provides this functionality without requiring explicit “back-edges” in the dependency graph.

Comment (superherointj) [reply pkgs/development/ocaml-modules/stdlib-shims/default.nix:14]: I have copied it from another example on nixpkgs. What do you want me to do? Remove all these tests?
Just let me know and I'll remove it.

File: pkgs/development/ocaml-modules/dose3/default.nix (hunk 5 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix]: I have disabled tests. Made an attempt of partial tests but didn't work.

File: pkgs/development/ocaml-modules/stdlib-shims/default.nix (hunk 6 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/stdlib-shims/default.nix]: I have removed all `passthru.tests` and clean-up `cudf`.
Enabling testing didn't go well, even partial.

**Comment by superherointj** [comment]
I was considering postponing this PR until tests are figured out but @anmonteiro has been succesfuly using Dose3 on his overlay:
https://github.com/anmonteiro/nix-overlays/blob/master/ocaml/dose3/default.nix

And it is very similar to what we have here. Despite tests being unavailable for now, the library can still prove useful (I hope).

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    parmap
+    re
+  ];
+  propagatedBuildInputs = [
+    bz2
+    camlzip
+    ocaml_extlib
+    ocamlgraph
+    stdlib-shims
+  ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dose3/default.nix:33]: How did you decide to propagate these and not the other ones?

Comment (vbgl) [reply pkgs/development/ocaml-modules/dose3/default.nix:31]: For instance, `parmap` must NOT be propagated, as it is only used for building the `dose-challenged` binary.

Comment (superherointj) [reply pkgs/development/ocaml-modules/dose3/default.nix:31]: Initially I was propagating everything. After you hinted to not do this, I started propagating nothing unless something downstream would error. The packages I have moved to propagated are the ones that showed up on downstream errors so far. I still haven't mastered the `ocamlfind` method to solve this problem. I honestly don't know in advance what to propagate or not.

**Pull request merged**